File: common/src/main/java/net/impleri/playerskills/utils/SkillResourceLocation.java
package net.impleri.playerskills.utils;

import net.minecraft.resources.ResourceLocation;

/**
 * Helper class for generating ResourceLocation, defaulting the namespace to playerSkills
 */
public class SkillResourceLocation {
    public static ResourceLocation of(ResourceLocation resourceLocation) {
        return resourceLocation;
    }

    public static ResourceLocation of(String namespace, String path) {
        return new ResourceLocation(namespace, path);
    }

    public static ResourceLocation of(String path) {
        String[] elements = decompose(path);

        return of(elements[0], elements[1]);
    }

    /**
     * ofMinecraft
     * <p>
     * Attempt to parse the string into a minecraft-default ResourceLocation and fallback to skills namespace if some error
     */
    public static ResourceLocation ofMinecraft(String value) {
        var mcResource = ResourceLocation.tryParse(value);

        return mcResource == null ? of(value) : mcResource;
    }

    private static final String DEFAULT_NAMESPACE = "skills";

    private static String[] decompose(String string) {
        String namespace = DEFAULT_NAMESPACE;
        String element = string;

        int i = string.indexOf(":");
        if (i >= 0) {
            element = string.substring(i + 1);

            if (i >= 1) {
                namespace = string.substring(0, i);
            }
        }

        return new String[]{namespace, element};
    }
}

File: common/src/main/kotlin/net/impleri/playerskills/utils/SkillResourceLocation.kt
package net.impleri.playerskills.utils

import net.minecraft.resources.ResourceLocation

/**
 * Helper class for generating ResourceLocation, defaulting the namespace to playerSkills
 */
object SkillResourceLocation {
  @JvmStatic
  fun of(resourceLocation: ResourceLocation): ResourceLocation {
    return resourceLocation
  }

  @JvmStatic
  fun of(namespace: String, path: String): ResourceLocation {
    return ResourceLocation(namespace, path)
  }

  @JvmStatic
  fun of(path: String): ResourceLocation {
    val elements = decompose(path)
    return of(elements[0], elements[1])
  }

  /**
   * ofMinecraft
   *
   *
   * Attempt to parse the string into a minecraft-default ResourceLocation and fallback to skills namespace if some error
   */
  @JvmStatic
  fun ofMinecraft(value: String): ResourceLocation {
    val mcResource = ResourceLocation.tryParse(value)
    return mcResource ?: of(value)
  }

  private const val DEFAULT_NAMESPACE = "skills"
  private fun decompose(string: String): Array<String> {
    var namespace = DEFAULT_NAMESPACE
    var element = string
    val i = string.indexOf(":")
    if (i >= 0) {
      element = string.substring(i + 1)
      if (i >= 1) {
        namespace = string.substring(0, i)
      }
    }
    return arrayOf(namespace, element)
  }
}
